File: Spawn-MPICH/master_src/master.cpp
#include <iostream>
#include <unistd.h>
#include <mpi.h>
#define MASTER 0

using namespace std;

//mpi
int rank;
int n_children;
MPI_Comm children;

char processor_name[MPI_MAX_PROCESSOR_NAME];
int name_len;

void master_work()
{
  std::cout << "[" << rank << "]" << "HELLO FROM MASTER @ " << processor_name << endl;
  MPI_Recv(&n_children, 1, MPI_INT, 0, 0, children, MPI_STATUS_IGNORE);
  std::cout << "[" << rank << "]" << " Got " << n_children << " slaves..." << endl;
  std::cout << "[" << rank << "]" << "BYE BYE FROM MASTER" << endl;
}

/**
* @method main
**/
int main(int argc, char* argv[])
{
  //init
  MPI::Init(argc,argv);
  rank = MPI::COMM_WORLD.Get_rank();
  MPI_Get_processor_name(processor_name, &name_len);

  //add hosts and spawn
  MPI_Info info;
  MPI_Info_create(&info);
  MPI_Info_set(info,"hostfile","./host_file");
  int spawn_callback = MPI_Comm_spawn(
	"./slave"/*name of program to be spawned*/,
	MPI_ARGV_NULL/*arguments to command*/,
	2/*max number of processes to start*/,
	info/*MPI_Info obj*/,
	0/*rank of process in which arguments are examined*/,
	MPI_COMM_WORLD/*MPI_Comm intracommunicator w/group of spawning proc*/,
	&children/*MPI_Comm intercommunicator between original group and new*/,
	MPI_ERRCODES_IGNORE/*array of err_codes*/);

  switch(spawn_callback)
  {
	case MPI_SUCCESS:
		std::cout << "spawn callback: MPI_SUCCESS" << endl;
		break;
	case MPI_ERR_COMM:
		std::cout << "spawn callback: MPI_ERR_COMM" << endl;
		break;
	case MPI_ERR_ARG:
		std::cout << "spawn callback: MPI_ERR_ARG" << endl;
		break;
	case MPI_ERR_INFO:
		std::cout << "spawn callback: MPI_ERR_INFO" << endl;
		break;
	case MPI_ERR_SPAWN:
		std::cout << "spawn callback: MPI_ERR_SPAWN" << endl;
		break;
  }

  //work
  master_work();
  sleep(3);

  //finialize
  MPI_Finalize();
  return 0;
}

File: spawn/java_master_/src/Master.java
import mpi.*;

class Master {
	
	//configuration
	static int maxprocs= 4;
	static int loops= 10;
	static String hostfile = "/mirror/romina/workspace/demo/mpi/spawn/host_file";
	static String command = "/mirror/romina/workspace/demo/mpi/spawn/slave";

	//spawn
	static int spawn() throws MPIException {		
		Info info= new Info();
		int[] errcode = new int[maxprocs];
		info.set("hostfile", hostfile);
		Intercomm children= MPI.COMM_WORLD.spawn(command,new String[]{},maxprocs,info,0,errcode);

		int n_children[] = new int [1];
		children.recv(n_children, 1, MPI.INT, 0, 0);

		children.free();
		return n_children[0];
	}

	public static void main(String[] args) throws MPIException {
		//init
		MPI.Init(args);
		//int rank = MPI.COMM_WORLD.getRank();
		//int size = MPI.COMM_WORLD.getSize();
		String name = MPI.getProcessorName();

		//work
		System.out.println("[MASTER] Hello @ "+name);
		for(int i=0; i<loops; i++)
			try {
				System.out.println("[MASTER] Spawn Loop#" + i + " Got " + spawn() +" slaves...");
			    	Thread.sleep(1000);
			} catch(Exception ex) {
			    	System.out.println("Excpetion happened");
			}

		//finalize
		System.out.println("[MASTER] Good Bye");
		MPI.Finalize();
	}

}
